Sources/CvCity.cpp

```diff
@@ -6521,61 +6509,49 @@ void CvCity::recalculateCultureDistances(int iMaxDistance) const
-	int iNewValue = 0;
 	// as long as there are changes during the last iteration
 	while (bHasChanged)
 	{
 		// reset the has changed variable to note a new loop cycle has begun
 		bHasChanged = false;
 
-		for (int iDX = -iMaxDistance; iDX <= iMaxDistance; iDX++)
+		foreach_(const CvPlot* plotX, plot()->rect(iMaxDistance, iMaxDistance))
 		{
-			for (int iDY = -iMaxDistance; iDY <= iMaxDistance; iDY++)
+			// Some tiles at center should not be calculated with formula (as they already are defined above):
+			if (plotDistance(getX(), getY(), plotX->getX(), plotX->getY()) < 2 &&
+				// Regular realistic culture, with any adjacency to city, and
+				(!GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING) ||
+				// 1-tile-start with cardinal adjacency to city
+				getX() == plotX->getX() || getY() == plotX->getY()))
 			{
-				// Some tiles at center should not be calculated with formula (as they already are defined above):
-				// 1-tile-start with cardinal adjacency to city, and
-				if ((GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING) && (abs(iDX) + abs(iDY) < 2)) ||
-					// regular realistic culture, with any adjacency to city
-					(!GC.getGame().isOption(GAMEOPTION_1_CITY_TILE_FOUNDING) && (abs(iDX) < 2 && abs(iDY) < 2)))
-				{
-					// This is a slightly cursed function.
-					continue;
-				}
-				plotIndex = HASH_RELATIVE_CLOSE_DIST(iDX, iDY);
-				// recalculate the value to determine if it has changed
-				iNewValue = calculateCultureDistance(iDX, iDY, iMaxDistance);
-
-				// if it has changed, save the value and mark that
-				//   ALL values (!) should be recomputed since they
-				//   may depend on this value
-				if (m_aCultureDistances[plotIndex] != iNewValue)
-				{
-					m_aCultureDistances[plotIndex] = iNewValue;
-					bHasChanged = true;
-				}
+				// This is a slightly cursed function.
+				continue;
+			}
+			// recalculate the value to determine if it has changed
+			const int iNewValue = calculateCultureDistance(plotX, iMaxDistance);
+
+			// if it has changed, save the value and mark that
+			//   ALL values (!) should be recomputed since they
+			//   may depend on this value
+			if (m_aCultureDistances[plotX] != iNewValue)
+			{
+				m_aCultureDistances[plotX] = iNewValue;
+				bHasChanged = true;
 			}
-		}
-		numLoops++;
-		if (numLoops > 100)
-		{
-			bHasChanged = false;
 		}
 	}
 }
 
-int CvCity::calculateCultureDistance(int iDX, int iDY, int iMaxDistance) const
+int CvCity::calculateCultureDistance(const CvPlot* mainPlot, int iMaxDistance) const
 {
 	// This function is the meat of Realistic Culture, determines extra distance/cost for tiles.
 	// Each point of distance corresponds 1:1 with level of city culture to reach it.
 	PROFILE_FUNC();
 
-	// find the current plot and the distance to it
-	const CvPlot* mainPlot = plotXY(getX(), getY(), iDX, iDY);
-
+	FAssertMsg(plotDistance(getX(), getY(), mainPlot->getX(), mainPlot->getY()) <= iMaxDistance, "test"); // Matt: does the following check ever pass?
```

Comment:
Not sure precisely the functionality of plot()->rect(iMaxDistance, iMaxDistance) as it calls this function, but if it effectively makes a rectangle of radius X, this check exists to avoid having to calculate needlessly the corners of said rectangle (relative coord X, X say) because our actual city's radius can't be larger than X itself.

Reply:
Next line will never happen right?

if (plotDistance(getX(), getY(), mainPlot->getX(), mainPlot->getY()) > iMaxDistance) return MAX_INT;

Reply:
I think it will frequently happen, unless I'm misunderstanding something.

Pretend iMaxDistance is 2 (regular big fat cross, 21 tiles). This loop to calculate the cultural tile distance from the city center will be called not on just the 21 tiles, but the full 25 tiles in the 5x5 rectangle. Given the plot distance of the corner tiles will be 3, greater than iMaxDistance of 2, we don't want to update/recalculate everything just to update those corners we won't need anyway, right?

Not much impact with max distance 2, but at max distance 10 with 20x20 tiles, given it basically loops thru the entire thing again everything when a single tile needs updating, removing the 50-some tiles seems like a prudent step.

Reply:
```cpp
// 4 | 4 | 3 | 3 | 3 | 4 | 4
// -------------------------
// 4 | 3 | 2 | 2 | 2 | 3 | 4
// -------------------------
// 3 | 2 | 1 | 1 | 1 | 2 | 3
// -------------------------
// 3 | 2 | 1 | 0 | 1 | 2 | 3
// -------------------------
// 3 | 2 | 1 | 1 | 1 | 2 | 3
// -------------------------
// 4 | 3 | 2 | 2 | 2 | 3 | 4
// -------------------------
// 4 | 4 | 3 | 3 | 3 | 4 | 4
//
// Returns the distance between plots according to the pattern above...
inline int plotDistance(int iX1, int iY1, int iX2, int iY2)```